### PR TITLE
feat: add option to draw labels on top of all `Polygon`s

### PR DESCRIPTION
**Pull request by ignatz**

### CODE REVIEW

File: lib/src/layer/polygon_layer/label.dart
Comment (ignatz) [inline lib/src/layer/polygon_layer/label.dart:23]: This check is super weird, it removes labels when their center goes off screen but only to the left. It's jarring to look at.

File: lib/src/layer/polygon_layer/polygon_layer.dart
Comment (ignatz) [inline lib/src/layer/polygon_layer/polygon_layer.dart:244]: I'd also be happy to just remove this entire block, i.e. the old behavior. I'm not sure what's the use case for drawing polygons over labels with labels sometimes sticking out partially. Or appearing faint when overdrawn by semi-transparent polygons.

File: lib/src/layer/polygon_layer/polygon_layer.dart
Comment (JaffaKetchup) [inline lib/src/layer/polygon_layer/polygon_layer.dart:294]: This `padding` is different from the `padding: 10` in the other `builtLabelTextPainter` call.

File: lib/src/layer/polygon_layer/polygon_layer.dart
Comment (JaffaKetchup) [reply lib/src/layer/polygon_layer/polygon_layer.dart:244]: Personally, I think the old behaviour might help with depth perception, so I think it would be a good idea to keep it.

File: lib/src/layer/polygon_layer/label.dart
Comment (JaffaKetchup) [reply lib/src/layer/polygon_layer/label.dart:23]: Wouldn't it be better to fix this to remove labels when off screen in any direction, rather than remove this optimization altogether like you appear to have done? (Unless I've missed something that is.)

Comment (ignatz) [reply lib/src/layer/polygon_layer/label.dart:23]: We could.

For now, I'd rather call it: "fixing visual glitches" rather than "removing an optimization that only 'work' when labels slip out to the left" :).

Computationally it would probably still be worth it. However, note that right now we're already culling polygons. So, this optimization would only kick in for polygons that are only partially on screen with the label off screen.

Would you wanna do it now or later?

File: lib/src/layer/polygon_layer/polygon_layer.dart
Comment (ignatz) [reply lib/src/layer/polygon_layer/polygon_layer.dart:294]: Good catch. It wasn't intentional.

I picked a bigger value for my app and 50 is what I settled on. Not sure if there's a single right value.

For context, the padding says at what point the label should go away when you zoom out. Specifically, the polygon gets smaller but the label keeps it size, so at some point when the label plus padding doesn't fully fit into the polygon anymore the code drops it. I found that a small padding will keep the label longer on screen but may make it look a bit crammed :shrug:

File: lib/src/layer/polygon_layer/polygon_layer.dart
Comment (ignatz) [reply lib/src/layer/polygon_layer/polygon_layer.dart:244]: Maybe. Maybe if the labels weren't text.

I found it to look like a car crash at best. At worst, it's an accessibility nightmare. Nothing better than a label you wanna read with lines and fills running over it :)

File: lib/src/layer/polygon_layer/polygon_layer.dart
Comment (JaffaKetchup) [reply lib/src/layer/polygon_layer/polygon_layer.dart:294]: Maybe we should go with something like 20 then, as I do remember seeing 10 as a little cramped.

File: lib/src/layer/polygon_layer/polygon_layer.dart
Comment (JaffaKetchup) [reply lib/src/layer/polygon_layer/polygon_layer.dart:244]: > At worst, it's an accessibility nightmare.

Absolutely true of course. Better to colour the text like GitHub does with it's issue labels.

But for the sake of not spoiling anyone's fun, let's just keep it.

File: lib/src/layer/polygon_layer/label.dart
Comment (JaffaKetchup) [reply lib/src/layer/polygon_layer/label.dart:23]: I guess that most of the time, the label is likely to be present. I guess technically 50% of the time if all locations had equal chance.
You're right, let's leave it for now. I think culling the polygons is probably enough.

File: lib/src/layer/polygon_layer/polygon_layer.dart
Comment (ignatz) [reply lib/src/layer/polygon_layer/polygon_layer.dart:294]: Happy to change, no strong feelings around any particular value.

Since you brought this up, I also remembered that I had bumped the "maxLines" from 1 to 2 for my use-case. Arguably the value is very use-case dependend, would you like me to add a configurable parameter or simply not limit the lines?

**Review by JaffaKetchup — APPROVED** [review]
LGTM, I'll wait on one more review.

**Review by TesteurManiak — APPROVED** [review]
LGTM!

**Pull request merged**
